### PR TITLE
gh-90633: Improve error and docs for typing.assert_never

### PR DESCRIPTION
**Pull request by JelleZijlstra**
#90633

### CODE REVIEW

**Review by Fidget-Spinner — APPROVED** [review]

**Review by AlexWaygood — COMMENTED** [review]
Perhaps it might be nice to include a one-sentence explanation of _how_ the type checker determines unreachability here (by narrowing the type such that the only satisfiable type is the empty bottom type, `Never`). That might clear up some of the user confusion that we had in the linked issue.

**Comment by JelleZijlstra** [comment]
@AlexWaygood good point, I expanded the docs.

File: Doc/library/typing.rst [hunk 1]
```diff
@@ -2341,11 +2341,22 @@ Functions and decorators
                case _ as unreachable:
                    assert_never(unreachable)
 
+   Here, the type annotations allow the type checker to infer that the
```
Comment (AlexWaygood) [inline Doc/library/typing.rst]: Too many types ;)

```suggestion
   Here, the annotations allow the type checker to infer that the
```

File: Doc/library/typing.rst [hunk 2]
```diff
@@ -2341,11 +2341,22 @@ Functions and decorators
...
+   last case can never execute, because ``arg`` is either
+   an :class:`int` or a :class:`str`, and both options are covered by
+   earlier cases.
    If a type checker finds that a call to ``assert_never()`` is
-   reachable, it will emit an error.
+   reachable, it will emit an error. For example, if the type annotation
+   for ``arg`` was instead ``int | str | float``, the type checker would
+   emit an error pointing out that ``unreachable`` is of type :class:`float`.
```
Comment (AlexWaygood) [inline Doc/library/typing.rst:2351]:
```suggestion
   emit an error pointing out that ``unreachable`` is of type :class:`float`.
   For a call to ``assert_never`` to succeed, the inferred type of
   the argument passed in must be the bottom type, :data:`Never`, and nothing
   else.
```

Comment (JelleZijlstra) [reply Doc/library/typing.rst:2351]: Thanks, applied both of your suggestions.

**Review by AlexWaygood — APPROVED** [review]

**Pull request merged**